src/Microsoft.OData.Core/ODataNotificationWriter.cs

```diff
@@ -324,13 +331,40 @@ protected override void Dispose(bool disposing)
                 if (this.listener != null)
                 {
                     // Tell the listener that the stream is being disposed.
-                    this.listener.StreamDisposed();
+                    if (synchronous)
+                    {
+                        this.listener.StreamDisposed();
+                    }
+                    else
+                    {
+                        this.listener.StreamDisposedAsync().Wait();
+                    }
+
                     this.listener = null;
                 }
             }
 
             this.textWriter.Dispose();

```

Comment:
By the way `TextWriter` also implements `IAsyncDisposable`. So maybe you could call `textWriter.Dispose()` in the `Dispose(bool)` method and `textWriter.DisposeAsync()` in the `DisposeAsync()` method.

Reply:
@habbes I'm not sure why but `.DisposeAsync()` method doesn't appear available to be invoked directly. However, `textWriter.Dispose()` should do the job since `TextWriter` is a BCL type.

Reply:
Interesting that it's not available on the class, yet it appears in the public API: https://docs.microsoft.com/en-us/dotnet/api/system.io.textwriter.disposeasync?view=net-5.0

Maybe we're reference an older version.

In that case I think it makes sense to call `this.textWriter?.Dispose()` outside the if-block. Alternatively, we can keep it in the if-block and duplicate it in both `Dispose()` and `DisposeAsync()`.